cuspatial/trajectory_clustering.ipynb

```diff
@@ -0,0 +1,720 @@
+    "\n",
+    "In this trajectory clustering example, we work on a particular vehicle trajectory dataset. Here are some summary statistics of the dataset.\n",
+    "\n",
+    "| num_trajectories | 10070 |\n",
+    "|------------------|-------|\n",
+    "| max_length       | 961   |\n",
+    "| min_length       | 30    |\n",
+    "| avg_length       | 87    |\n",
+    "\n",
+    "We use Hausdorff distance as the similarity metric between trajectories. We compute the pairwise similarity matrix and apply AgglomerativeClustering (AC) and DBSCAN afterwards.\n",
+    "\n",
+    "In this example, cuSpatial can significantly accelerate the computation of the Hausdorff distance similarity matrix. Comparing to the typical scikit-learn implementation of Hausdorff distance (single CPU thread), cuSpatial reduces the computation time from about 5 hours to under 15 seconds on this particular dataset.\n",
+    "\n",
+    "Since we pre-computed the similarity matrix, we can much more easily experiment with different clustering algorithm and different hyperparameter sets. In this example we do it in an interactive way using ipython widgets, you may play with the hyperparameter set and see how the clustering result responds.\n",
+    "\n",
+    "## Install Dependencies\n",
+    "\t\n",
+    "1. cuSpatial:    `conda install -c rapidsai-nightly cuspatial`\n",
+    "2. scikit-image: `conda install -c conda-forge scikit-image`\n",
+    "3. opencv: \t     `conda install -c conda-forge opencv`\n",
+    "4. ipywidgets: \t `conda install -c anaconda ipywidgets`\n",
+    "5. scikit-learn: `conda install -c anaconda scikit-learn`\n",
+    "\n",
+    "More [instructions on installing cuSpatial](https://github.com/rapidsai/cuspatial) (e.g. building from source)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from skimage import io, transform\n",
+    "import numpy as np\n",
+    "import matplotlib.pyplot as plt\n",
+    "import cv2\n",
+    "import os\n",
+    "import random\n",
+    "from ipywidgets import interact, interactive, fixed\n",
+    "from ipywidgets import *\n",
+    "import time\n",
+    "\n",
+    "from scipy.spatial.distance import directed_hausdorff\n",
+    "from sklearn.cluster import AgglomerativeClustering,KMeans,DBSCAN\n",
+    "\n",
+    "import cudf\n",
+    "import cuspatial\n",
+    "from matplotlib import gridspec\n",
+    "\n",
+    "def hausforff_distance(u,v):\n",
+    "    return max(directed_hausdorff(u, v)[0], directed_hausdorff(v, u)[0])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load and prepare the data"
```

Comment:
```suggestion
    "## Load and prepare the data"
    
    For comparison's sake, we'll first implement the CPU-only version of this analysis.
```

Perhaps you want to use different verbiage, but I had to scroll up and down a few times to confirm that the first several cells are host-only code.

It helps to describe step by step what's about to happen before providing the code.

Reply:
A numbered list of what this notebook will do helps users follow along. Something like..

"We'll compare CPU only vs GPU accelerated implementations.

For the CPU version, we need to:
1. Load the data from disk
2. Define trajectory calculation functions
3. Filter out noise and apply the trajectory calculations
4. Do a visual inspection of the computed trajectories
5. Cluster the trajectories using Hausdorff distance as the similarity metric

...
much code
...

For the GPU accelerated implementation, we implement the same logic with mostly builtin functions:
1. foo
2. bar
3. profit